### PR TITLE
[ML] Validate classification dependent_variable cardinality is at lea…

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
…st two

Data frame analytics classification currently only supports 2 classes for the
dependent variable. We were checking that the field's cardinality is not higher
than 2 but we should also check it is not less than that as otherwise the process
fails.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/DataFrameAnalysis.java [hunk 1]
```diff
@@ -37,9 +37,9 @@
     List<RequiredField> getRequiredFields();
 
     /**
-     * @return {@link Map} containing cardinality limits for the selected (analysis-specific) fields
+     * @return {@link List} containing cardinality limits for the selected (analysis-specific) fields
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/DataFrameAnalysis.java]: s/limits/constraints
?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/DataFrameAnalysis.java [hunk 2]
```diff
@@ -37,9 +37,9 @@
...
      */
-    Map<String, Long> getFieldCardinalityLimits();
+    List<FieldCardinalityConstraint> getFieldCardinalityLimits();
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/DataFrameAnalysis.java]: s/Limits/Constraints
?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/OutlierDetectionTests.java [hunk 3]
```diff
@@ -89,7 +89,7 @@ public void testRequiredFieldsIsEmpty() {
     }
 
     public void testFieldCardinalityLimitsIsEmpty() {
-        assertThat(createTestInstance().getFieldCardinalityLimits(), is(anEmptyMap()));
+        assertThat(createTestInstance().getFieldCardinalityLimits().isEmpty(), is(true));
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/OutlierDetectionTests.java]: There is a "Matchers.empty()" matcher that could be used here.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/RegressionTests.java [hunk 4]
```diff
@@ -106,7 +106,7 @@ public void testRequiredFieldsIsNonEmpty() {
     }
 
     public void testFieldCardinalityLimitsIsEmpty() {
-        assertThat(createTestInstance().getFieldCardinalityLimits(), is(anEmptyMap()));
+        assertThat(createTestInstance().getFieldCardinalityLimits().isEmpty(), is(true));
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/RegressionTests.java]: There is a "Matchers.empty()" matcher that could be used here.

**Review by przemekwitek — APPROVED** [review]
LGTM

Just a few minor comments

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsManager.java [hunk 5]
```diff
@@ -177,6 +177,7 @@ private void reindexDataframeAndStartAnalysis(DataFrameAnalyticsTask task, DataF
         ActionListener<CreateIndexResponse> copyIndexCreatedListener = ActionListener.wrap(
             createIndexResponse -> {
                 ReindexRequest reindexRequest = new ReindexRequest();
+                reindexRequest.setRefresh(true);
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsManager.java]: Note that this is now necessary as we are checking the field cardinality before we call `startAnalytics` which refreshes the dest index.

**Comment by dimitris-athanasiou** [comment]
@przemekwitek I have addressed all your points plus fixes a bug regarding refreshing of the dest index which was caught by the tests.

File: x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/ClassificationIT.java [hunk 6]
```diff
@@ -492,7 +492,7 @@ private void initialize(String jobId) {
         this.jobId = jobId;
         this.sourceIndex = jobId + "_source_index";
         this.destIndex = sourceIndex + "_results";
-        this.analysisUsesExistingDestIndex = randomBoolean();
+        this.analysisUsesExistingDestIndex = false;
```
Comment (przemekwitek) [inline x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/ClassificationIT.java]: Should this be reverted?

**Review by przemekwitek — APPROVED** [review]

**Comment by przemekwitek** [comment]
> @przemekwitek I have addressed all your points plus fixes a bug regarding refreshing of the dest index which was caught by the tests.

Please see my comment in `ClassificationIT`. Otherwise the PR is good to go.

File: x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/ClassificationIT.java (hunk 6 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/ClassificationIT.java]: Ah, yes of course.

**Pull request merged**